crates/ra_ide/src/hover.rs

```diff
@@ -18,14 +18,64 @@ use ra_syntax::{
 };
 
 use crate::{
-    display::{macro_label, rust_code_markup, rust_code_markup_with_doc, ShortLabel},
-    FilePosition, RangeInfo,
+    display::{macro_label, rust_code_markup, rust_code_markup_with_doc, ShortLabel, ToNav},
+    runnables::runnable,
+    FileId, FilePosition, NavigationTarget, RangeInfo, Runnable,
 };
 
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct HoverConfig {
+    pub run: bool,
+    pub debug: bool,
+    pub goto_type_def: bool,
+    pub implementations: bool,
+}
+
+impl Default for HoverConfig {
+    fn default() -> Self {
+        // A client should explicitly specify if it supports hover actions!
+        Self { run: false, debug: false, goto_type_def: false, implementations: false }
+    }
+}
+
+impl HoverConfig {
+    pub fn runnable(&self) -> bool {
+        self.run || self.debug
+    }
+
+    pub fn any(&self) -> bool {
+        self.implementations || self.goto_type_def || self.runnable()
+    }
+
+    pub fn none(&self) -> bool {
+        !self.any()
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct HoverGotoTypeData {
+    pub hint: String,
```

Comment:
the meaning of `hint` is not immediately obvious to me, so a comment (or maybe a rename) is wanted.

Reply:
It's a tooltip for the command link, when represented in the UI. I'll rename.

Reply:
```suggestion
    pub tooltip: String,
```